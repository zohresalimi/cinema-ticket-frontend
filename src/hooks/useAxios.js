import { useEffect, useReducer, useMemo, useCallback } from "react";
import axios from "axios";
import qs from "qs";

const baseUrl = process.env.REACT_APP_BASE_URL;
const actions = {
  REQUEST_START: "REQUEST_START",
  REQUEST_END: "REQUEST_END",
};

async function fetchData({ path, ...config }, dispatch) {
  try {
    dispatch({ type: actions.REQUEST_START });
    const result = await axios({
      ...config,
      url: `${baseUrl}${path}`,
      paramsSerializer: (params) => {
        return qs.stringify(params);
      },
    });
    dispatch({ type: actions.REQUEST_END, payload: result.data });
  } catch (err) {
    dispatch({
      type: actions.REQUEST_END,
      payload: err,
      error: true,
    });
  }
}

const useAxios = (path, config) => {
  function reducer(state, action) {
    switch (action.type) {
      case actions.REQUEST_START:
        return {
          ...state,
          response: null,
          loading: true,
          error: null,
        };
      case actions.REQUEST_END:
        return {
          ...state,
          loading: false,
          ...(action.error ? {} : { data: action.payload.data }),
          [action.error ? "error" : "response"]: action.payload,
        };
      default:
        return state;
    }
  }

  function createInitialState({ manual }) {
    return {
      loading: !manual,
      error: null,
      response: null,
    };
  }

  const options = useMemo(
    () => ({
      data: {},
      method: "get",
      manual: false,
      path,
      ...config,
    }),
    [config, path]
  );
  const [state, dispatch] = useReducer(reducer, createInitialState(options));

  useEffect(() => {
    if (!options.manual) {
      fetchData(options, dispatch);
    }
  }, [options, dispatch, path]);

  const refetch = useCallback(
    (payload) => {
      return fetchData({ ...options, ...payload }, dispatch);
    },
    [options, dispatch]
  );

  return [state, refetch];
};

export default useAxios;
