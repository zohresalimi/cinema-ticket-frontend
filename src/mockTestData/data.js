import React, { useReducer } from "react";
import { Router } from "@reach/router";
import AppContext from "../store/context";
import Theme from "../Styles/Theme";
import reducer from "../store/reducer";

export const getTestStore = () => {
  return {
    user: {
      name: "",
      email: "",
    },
    movies: {
      children: [
        {
          _id: "5fa71823adc463397a50c501",
          actors: ["Holland Roden", "Ronen Rubinstein", "Keegan Allen"],
          age: "7",
          coverImage:
            "https://catalog.cinema-api.com/cf/images/ncg-images/68bfeef55ada4e7ab0606db2af0f40ad.jpg?width=240&version=EA2E80EE161E5DA161C6E26487054C9F&format=webp",
          description:
            "Poppy and Kvist embark on an adventure that will take them far beyond",
          director: "aaron woodley",
          duration: "1:31",
          genre: ["family", "children"],
          images: [],
          largeImage:
            "https://catalog.cinema-api.com/cf/images/ncg-images/79276cead82341d2a004f8e8f99a88e9.jpg?width=1920&version=08F0C83177D59FB94C4C809F63F67E2A&format=webp",
          name: "trolls 2: world tour",
          originalTitle: "swedish",
          premiere: "2020-10-03T15:18:52.722Z",
          rooms: ["5fa653d0a7f1201999dd1ac0"],
          trailerUrl: "https://youtu.be/P9A6hOg9QQ0",
        },
      ],
      premiered: [
        {
          _id: "5fa71823adc463397a50c502",
          actors: ["Holland Roden", "Ronen Rubinstein", "Keegan Allen"],
          age: "7",
          coverImage:
            "https://catalog.cinema-api.com/cf/images/ncg-images/68bfeef55ada4e7ab0606db2af0f40ad.jpg?width=240&version=EA2E80EE161E5DA161C6E26487054C9F&format=webp",
          description:
            "Poppy and Kvist embark on an adventure that will take them far beyond",
          director: "aaron woodley",
          duration: "1:31",
          genre: ["family", "children"],
          images: [],
          largeImage:
            "https://catalog.cinema-api.com/cf/images/ncg-images/79276cead82341d2a004f8e8f99a88e9.jpg?width=1920&version=08F0C83177D59FB94C4C809F63F67E2A&format=webp",
          name: "follow me",
          originalTitle: "swedish",
          premiere: "2020-10-05T15:18:52.722Z",
          rooms: ["5fa653c0643c06586bba7d78"],
          trailerUrl: "https://youtu.be/P9A6hOg9QQ0",
        },
      ],
      upcoming: [
        {
          _id: "5fa71823adc463397a50c503",
          actors: ["Holland Roden", "Ronen Rubinstein", "Keegan Allen"],
          age: "7",
          coverImage:
            "https://catalog.cinema-api.com/cf/images/ncg-images/68bfeef55ada4e7ab0606db2af0f40ad.jpg?width=240&version=EA2E80EE161E5DA161C6E26487054C9F&format=webp",
          description:
            "Poppy and Kvist embark on an adventure that will take them far beyond",
          director: "aaron woodley",
          duration: "1:31",
          genre: ["family", "children"],
          images: [],
          largeImage:
            "https://catalog.cinema-api.com/cf/images/ncg-images/79276cead82341d2a004f8e8f99a88e9.jpg?width=1920&version=08F0C83177D59FB94C4C809F63F67E2A&format=webp",
          name: "honest thief",
          originalTitle: "swedish",
          premiere: "2021-01-03T15:18:52.722Z",
          rooms: ["5fa653c0643c06586bba7d78"],
          trailerUrl: "https://youtu.be/P9A6hOg9QQ0",
        },
      ],
      allMovies: {
        "5fa71823adc463397a50c501": "children",
        "5fa71823adc463397a50c502": "premiered",
        "5fa71823adc463397a50c503": "upcoming",
      },
    },
    cinemas: [
      {
        _id: "5fa64fe234fbbf9945f3a3ce",
        name: "filmstaden kista",
        purchaseEndTime: "21:30",
        purchaseStartTime: "8:00",
        rooms: ["5fa653c0643c06586bba7d78", "5fa653d0a7f1201999dd1ac0"],
      },
    ],
    ticket: {
      movie: {},
      status: false,
      quantity: 0,
      price: 0,
      seatNumbers: new Map(),
      user: {},
      showing: {},
    },
    showings: [
      {
        _id: "5fbe9bd31f58462b5c759b5c",
        capacity: 4,
        cinema: "5fa64fe234fbbf9945f3a3ce",
        endTime: null,
        movie: {
          _id: "5fa71823adc463397a50c502",
          actors: ["Holland Roden", "Ronen Rubinstein", "Keegan Allen"],
          age: "7",
          description:
            "Poppy and Kvist embark on an adventure that will take them far beyond",
          director: "aaron woodley",
          duration: "1:31",
          genre: ["family", "children"],
          images: [],
          name: "follow me",
          originalTitle: "swedish",
          premiere: "2020-10-05T15:18:52.722Z",
          rooms: ["5fa653c0643c06586bba7d78"],
          trailerUrl: "https://youtu.be/P9A6hOg9QQ0",
        },
        price: 100,
        room: { _id: "5fa653c0643c06586bba7d78" },

        seats: [
          [
            { seatNumber: 1, taken: "available" },
            { seatNumber: 2, taken: "available" },
          ],
          [
            { seatNumber: 1, taken: "available" },
            { seatNumber: 2, taken: "available" },
          ],
        ],
        startTime: "2020-11-13T16:00:00.000Z",
      },
    ],
  };
};

export const WithProvider = (props) => {
  const [state, dispatch] = useReducer(reducer, getTestStore());

  return (
    <Theme>
      <AppContext.Provider
        value={{
          state: props.defaultValue || state,
          dispatch: props.mockDispatch || dispatch,
        }}
      >
        <Router>{props.children}</Router>
      </AppContext.Provider>
    </Theme>
  );
};
