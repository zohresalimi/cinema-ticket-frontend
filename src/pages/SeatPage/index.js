/* eslint-disable no-restricted-syntax */
/* eslint-disable jsx-a11y/click-events-have-key-events */
/* eslint-disable jsx-a11y/control-has-associated-label */
import React, { useContext, useEffect, useState } from "react";
import AppContext from "../../store/context";
import { SET_SELECTED_SEAT } from "../../constants";

function SeatPlan({ seats }) {
  const { state, dispatch } = useContext(AppContext);
  const { quantity, seatNumbers } = state.ticket;
  const [ticketCount, setTicketCount] = useState(0);

  const bookSeat = (rowIndex, colIndex) => {
    dispatch({ type: SET_SELECTED_SEAT, data: [rowIndex, colIndex] });
  };

  useEffect(() => {
    let count = 0;
    for (const [_, columns] of seatNumbers) {
      count += columns.size;
    }
    setTicketCount(count);
  }, [seatNumbers, setTicketCount]);

  // const isDisabled = () => {
  //   if (quantity > seatNumbers.length) {
  //     seatNumbers.showing.seat.map((seat) => !seat.taken);
  //   }
  // };

  return (
    <div className="rowrapper">
      {`${ticketCount}/${quantity}`}
      {seats.map((row, rowIndex) => (
        <div className="column-wrapper" key={`row-${rowIndex}`}>
          {row.map((column, colIndex) => {
            return (
              // eslint-disable-next-line jsx-a11y/interactive-supports-focus
              <div
                role="button"
                disabled={quantity === ticketCount && !column.taken}
                className={"seat " + (column.taken ? "taken" : "")}
                key={`column-${column.seatNumber}`}
                onClick={() => bookSeat(rowIndex, colIndex)}
              />
            );
          })}
        </div>
      ))}
    </div>
  );
}

function Seat() {
  const { state } = useContext(AppContext);
  const { showing } = state.ticket;

  return (
    <div>
      {showing && <SeatPlan seats={showing.seats} />}
      <button>continue</button>
    </div>
  );
}

export default Seat;
