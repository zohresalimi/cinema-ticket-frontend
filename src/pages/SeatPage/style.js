import styled from "styled-components";

const Wrapper = styled.div`
  font-family: ${(props) => props.theme.fontFamily.sf};
  width: 100%;
  margin-top: 40px;
  font-size: ${(props) => props.theme.fontSizes.content};
  color: ${(props) => props.theme.colors.textColor};
  margin-bottom: 20px;
  display: flex;
  flex-direction: column;
  .guideline {
    display: flex;
    flex-direction: row;
    justify-content: center;
    align-items: center;
    border-top: 1px solid #ccc;
    margin: 20px 0;
    padding: 20px 0;
    svg {
      width: 30px;
      margin: 0 10px;
    }

    .available {
      svg {
        fill: #ffffff;
        stroke: #797979;
        stroke-width: 8px;
      }
    }
    .taken {
      svg {
        fill: #636363;
      }
    }

    .reserved {
      svg {
        pointer-events: none;
        animation: blinker 1s linear infinite;
        fill: #ecc31a;
        stroke: #cca505;
      }
    }
    .selected {
      svg {
        fill: rgb(235 51 87);
      }
    }
  }
  justify-content: center;

  p {
    font-size: ${(props) => props.theme.fontSizes.small};
    color: ${(props) => props.theme.colors.darkGray};
    text-align: center;
    span {
      font-size: ${(props) => props.theme.fontSizes.medium};
      color: ${(props) => props.theme.colors.black};
      margin: 0 10px;
    }
  }
  .column-wrapper {
    display: flex;
    flex-direction: row;
    justify-content: center;

    div[disabled] {
      svg {
        fill: #ccc;
        stroke: #797979;
      }
      pointer-events: none;
    }

    .seat {
      width: 30px;
      height: 30px;
      margin: 3px;
      cursor: pointer;
      fill: #ffffff;
      stroke: #797979;
      stroke-width: 8px;
      &.taken {
        fill: #636363;
        &::before {
          content: "x";
          color: #454343;
          font-weight: 300;
          font-family: Arial, sans-serif;
          text-align: center;
          display: block;
          font-size: 25px;
        }
      }
      &.reserved {
        pointer-events: none;
        animation: blinker 1s linear infinite;
      }
      a:hover {
        fill: "green";
      }
      &.selected {
        fill: rgb(235 51 87);
      }
      &.reserved {
        fill: #ecc31a;
        stroke: #cca505;
        stroke-width: 8px;
      }
    }
  }

  @keyframes blinker {
    50% {
      opacity: 0.5;
    }
  }
`;

export { Wrapper as default };
