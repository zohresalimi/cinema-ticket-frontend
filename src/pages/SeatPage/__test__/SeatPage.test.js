import * as React from "react";
import * as ReachRouter from "@reach/router";

import { act, render } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";

import * as hooks from "../../../hooks/useAxios";
import { getTestStore, WithProvider } from "../../../mockTestData/data";
import SeatPage from "../";

jest.mock("react-i18next", () => ({
  useTranslation: () => ({
    t: (k) => k,
    i18n: {
      language: "sv",
      changeLanguage: () => new Promise(() => {}),
    },
  }),
}));

async function renderWrapper(setDefaultValue, props) {
  let component;
  const testState = getTestStore();
  const defaultValue = setDefaultValue
    ? {
        ...testState,
        ticket: {
          ...testState.ticket,
          showing: testState.showings[0],
        },
      }
    : null;

  act(() => {
    component = render(
      <WithProvider defaultValue={defaultValue}>
        <SeatPage {...props} path="/" />
      </WithProvider>
    );
  });

  return component;
}

describe("Seat Page Component Testing", () => {
  afterAll(() => {
    jest.clearAllMocks();
  });

  test("take snapshot", async () => {
    const mockUseLocation = jest.fn(() => ({
      state: {
        category: "premiered",
      },
    }));
    jest.spyOn(ReachRouter, "useLocation").mockImplementation(mockUseLocation);
    const { container } = await renderWrapper(true);
    expect(container.firstChild).toMatchSnapshot();
  });
});
