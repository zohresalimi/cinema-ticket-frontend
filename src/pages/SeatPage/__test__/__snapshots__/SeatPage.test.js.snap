// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Seat Page Component Testing take snapshot 1`] = `
<div
  style="outline: none;"
  tabindex="-1"
>
  <div
    class="StyleComponents__Container-sc-11erdxt-0 fFgwWO"
  >
    <div
      class="StyleComponents__Row-sc-11erdxt-1 jhoSVF"
    >
      <div
        class="full-width"
      >
        <h2
          class="StyleComponents__Breadcrumb-sc-11erdxt-4 cNjPuO"
        >
          <a
            href="/booking/undefined"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-angle-left fa-w-8 mr-10"
              data-icon="angle-left"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 256 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M31.7 239l136-136c9.4-9.4 24.6-9.4 33.9 0l22.6 22.6c9.4 9.4 9.4 24.6 0 33.9L127.9 256l96.4 96.4c9.4 9.4 9.4 24.6 0 33.9L201.7 409c-9.4 9.4-24.6 9.4-33.9 0l-136-136c-9.5-9.4-9.5-24.6-.1-34z"
                fill="currentColor"
              />
            </svg>
            ticket selection
          </a>
        </h2>
        <div
          class="style__Wrapper-gzbnn9-0 hhfVwl"
        >
          <p>
            selected
            <span>
              0
            </span>
            from
            <span>
              0
            </span>
          </p>
          <div
            class="column-wrapper"
          >
            <div
              class="seat available"
              disabled=""
              role="button"
            />
            <div
              class="seat available"
              disabled=""
              role="button"
            />
          </div>
          <div
            class="column-wrapper"
          >
            <div
              class="seat available"
              disabled=""
              role="button"
            />
            <div
              class="seat available"
              disabled=""
              role="button"
            />
          </div>
        </div>
        <button
          class="StyleComponents__Button-sc-11erdxt-3 kZtNg"
        >
          continue
        </button>
        <p>
          you must select at least one ticket
        </p>
      </div>
    </div>
  </div>
</div>
`;
